### PR TITLE
[IMP] website: Reduce crawer tests execution time

### PR DESCRIPTION
**Pull request by tivisse**
Skip url duplicates based on the URL path only (get rid of the query parameters)

Takes 30 seconds to execute all the crawler tests
image
Instead of 10 minutes in master
image
Could be backported in 15.0 as well, where it takes more than 20 minutes.
image

cc @Xavier-Do @d-fence @rdeodoo

### CODE REVIEW

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Very nice :+1:
It was still in my pipe but good thing you took care of it.

Could you visit the X first url-type instead of only the very first? I guess 3 would be fine.
It would not add much time but would be a quick win to test most flows. Indeed most of the time we setup differently the first demo data, for instance:
- Events: The first one has all the menu (and so pages) enabled, while the second doesn't
- Shop: The first (Customizable Desk) has generally different options set from the other products
- Forum: The first has tag pages set while the second doesn't
- Slides: The third one has a very different layout config than the first two, rendering a whole different view stack
- Portal invoices: First one is paid, second one is not, more button / different layout
- Portal tasks: First one has a lot of timesheet while second doesn't
- Maybe other modules too, and surely sub-pages of modules too differ from first/second/third record.

**Comment by Xavier-Do** [comment]
If website approves it is a big +1 for me.

Suggestion, shouldn't we keep  a version with params in a nightly build or it isn't the purpose of the test?

**Comment by rdeodoo** [comment]
> Suggestion, shouldn't we keep a version with params in a nightly build or it isn't the purpose of the test?

Would be great indeed, and you could add it in our own website nightly testing suite so we are notified when it's red :+1:

**Comment by tivisse** [comment]
Hi @rdeodoo 

Thanks for the review. :)

I've applied your suggestion, and we got an execution time of 3.5 minutes. Summary:
- All URLs: 10 minutes
- 3 URLs with same path max: 3.5 minutes
- 1 URL of same path max: 30 seconds

I let you check if there my implementation is valid or if I missed something. 

image

**Comment by rdeodoo** [comment]
Thanks for the changes, I had a quick looks at all the runbot logs and I believe there is a misunderstanding of what we want to achieve.
During the meeting, if I am not wrong, AL suggested to only visit the first URL we find for a given format/slug (eg `/blog/<blog.blog>/<blog.post>`, instead of all of them like it is now.
Then I replied that it would be nice to at least keep visiting the 3 first different URL (arbitrary but better than 1?) we would find since those generally test different setup/things (as I tried to explain [here](https://github.com/odoo/odoo/pull/146767#pullrequestreview-1788525416)) but I may have wrongly express what I meant since it was not understood that way, it's probably my bad, I should have illustrated with concrete examples.

GET /blog/<blog_blog>/<blog_post>
- master: 24 queries (all 8 blog posts (spread between 2 blogs) visited for each of public/demo/admin)
- Your V2 (after my review): same as master
- Your V1 (before my review): same as master
- My idea was: 3x3 queries (admin/demo/public): /blog/x/1 /blog/x/2 /blog/x/3

GET /my/purchase/
- master: 28 queries (14 for demo, 14 for admin, mix of ID (9, 10, 16, 17..) + mix of no query / &report_type=pdf )
- Your V2 (after my review): same as master
- Your V1 (before my review): 2 queries on /my/purchase/4?access_tok.. (one for demo and one for admin)
- My idea was: 6 or 12 queries: /my/purchase/x, x+1 and x+2 (both for admin and demo), + maybe same with &report_type=pdf

GET /my/tasks/
- master: 2077 queries
- Your V2 (after my review): 253 queries
- Your V1 (before my review): 2 queries: /my/tasks/65?acc.. for demo and /my/tasks/57 for admin
- My idea was: 6 or 12 queries: /my/tasks/x, x+1 and x+2, + maybe same with different query strings? (Once for admin and once for demo)

tl;dr:
- Visit /my/tasks/1, /my/tasks/2, /my/tasks/3 but not any other /my/tasks/<slug/id>
- For blog, visit:
  - /blog/travel-1/sierra-tarahumara-1
  - /blog/tag/adventure-2
  - /blog/travel-1/how-to-choose-the-right-hotel-3
  
  But not the following onces since we already got 3 :point_up: :
  - /blog/astronomy-2/how-to-look-up-4
  - /blog/astronomy-2/what-if-they-let-you-run-the-hubble-5
  - /blog/astronomy-2/buying-a-telescope-6
  - /blog/astronomy-2/beyond-the-eye-7

May I have your opinion about this? My rationale is that I have no clue how good are those tests against introducing code that would break feature and/or pages, since if it breaks people can't merge and have to fix it before we see it (purpose of a test).
I am thus afraid to remove all of those and keep only one. It's why I suggested 3, especially since in almost all apps the first 3 are using different setup/layouts, but I am maybe going to far :shrug: 

-----------------------------

Also, about @Xavier-Do remarks, could you add a new test class calling those crawler without any restriction flagged for nightly? Maybe just add `@tagged('website_nightly', '-standard')` to that new class, it should land in our own nightly website build so we are warned when it fails (I'll triple check it's enough).
